src/gt4py/next/iterator/transforms/infer_domain.py

```diff
@@ -0,0 +1,236 @@
+# GT4Py - GridTools Framework
+#
+# Copyright (c) 2014-2023, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import dataclasses
+
+from gt4py.eve.extended_typing import Dict, Tuple, Union
+from gt4py.next import common
+from gt4py.next.common import Dimension
+from gt4py.next.iterator import ir as itir
+from gt4py.next.iterator.ir_utils import ir_makers as im
+from gt4py.next.iterator.transforms.global_tmps import SymbolicDomain, SymbolicRange, domain_union, \
+    _max_domain_sizes_by_location_type
+from gt4py.next.iterator.transforms.trace_shifts import TraceShifts
+
+
+@dataclasses.dataclass(frozen=True)
+class InferDomain:
```

Comment:
Let's remove this class (at least for now) and make all class method free functions. There appears to be no need to have a class.